File: apps/api/app/models/market_enrichment.py
"""Market Data Enrichment models — DataSource registry, fetch logs, raw/processed data, review queue."""

from __future__ import annotations

import uuid
from datetime import date, datetime
from decimal import Decimal

from sqlalchemy import (
    Boolean,
    Date,
    DateTime,
    ForeignKey,
    Index,
    Integer,
    Numeric,
    String,
    Text,
    UniqueConstraint,
)
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.models.base import BaseModel


class MarketDataSource(BaseModel):
    """Registry of data sources with legal_basis tracking and tier classification."""

    __tablename__ = "market_data_sources"

    org_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False)
    name: Mapped[str] = mapped_column(String(200), nullable=False)
    slug: Mapped[str] = mapped_column(String(100), nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    source_type: Mapped[str] = mapped_column(
        String(30), nullable=False
    )  # "official_api" | "rss_feed" | "document" | "manual"
    tier: Mapped[int] = mapped_column(
        Integer, nullable=False
    )  # 1=Official APIs, 2=RSS, 3=Documents, 4=Manual
    base_url: Mapped[str | None] = mapped_column(String(500), nullable=True)
    legal_basis: Mapped[str] = mapped_column(
        String(50), nullable=False, server_default="public_data"
    )  # "public_data" | "licensed" | "fair_use" | "manual_entry"
    rate_limit_per_hour: Mapped[int] = mapped_column(Integer, default=60, server_default="60")
    is_active: Mapped[bool] = mapped_column(Boolean, default=True, server_default="true")
    config: Mapped[dict] = mapped_column(JSONB, default=dict, server_default="{}")
    created_by: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id", ondelete="SET NULL"), nullable=True
    )

    __table_args__ = (
        UniqueConstraint("org_id", "slug", name="uq_market_data_source_slug"),
        Index("ix_market_data_sources_org_id", "org_id"),
        Index("ix_market_data_sources_tier", "tier"),
        Index("ix_market_data_sources_is_active", "is_active"),
    )


class MarketEnrichmentFetchLog(BaseModel):
    """Log of each data fetch attempt — status, record counts, errors."""

    __tablename__ = "market_enrichment_fetch_logs"

    org_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False)
    source_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("market_data_sources.id", ondelete="CASCADE"),
        nullable=False,
    )
    status: Mapped[str] = mapped_column(
        String(20), nullable=False, server_default="pending"
    )  # "pending" | "running" | "success" | "failed" | "rate_limited"
    records_fetched: Mapped[int] = mapped_column(Integer, default=0, server_default="0")
    records_new: Mapped[int] = mapped_column(Integer, default=0, server_default="0")
    records_updated: Mapped[int] = mapped_column(Integer, default=0, server_default="0")
    error_message: Mapped[str | None] = mapped_column(String(2000), nullable=True)
    started_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    completed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

    __table_args__ = (
        Index("ix_enrichment_fetch_log_org", "org_id"),
        Index("ix_enrichment_fetch_log_source", "source_id"),
        Index("ix_enrichment_fetch_log_status", "status"),
    )


class MarketDataRaw(BaseModel):
    """Raw payload from each data source — stored verbatim for auditability."""

    __tablename__ = "market_data_raw"

    org_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False)
    source_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("market_data_sources.id", ondelete="CASCADE"),
        nullable=False,
    )
    fetch_log_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("market_enrichment_fetch_logs.id", ondelete="SET NULL"),
        nullable=True,
    )
    external_id: Mapped[str | None] = mapped_column(String(500), nullable=True)
    raw_content: Mapped[dict] = mapped_column(JSONB, nullable=False)
    content_hash: Mapped[str] = mapped_column(String(64), nullable=False)  # SHA-256
    fetched_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)

    __table_args__ = (
        UniqueConstraint("org_id", "source_id", "content_hash", name="uq_market_data_raw_dedup"),
        Index("ix_market_data_raw_content_hash", "content_hash"),
        Index("ix_market_data_raw_source", "source_id"),
        Index("ix_market_data_raw_org", "org_id"),
    )


class MarketDataProcessed(BaseModel):
    """Structured data extracted from raw records — typed, unit-aware, reviewable."""

    __tablename__ = "market_data_processed"

    org_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False)
    raw_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("market_data_raw.id", ondelete="SET NULL"),
        nullable=True,
    )
    data_type: Mapped[str] = mapped_column(
        String(50), nullable=False
    )  # "price" | "policy" | "project_pipeline" | "macro_indicator" | "news"
    category: Mapped[str] = mapped_column(
        String(100), nullable=False
    )  # e.g. "solar_ppi", "capacity_factor", "feed_in_tariff"
    region: Mapped[str | None] = mapped_column(String(100), nullable=True)
    technology: Mapped[str | None] = mapped_column(String(100), nullable=True)
    effective_date: Mapped[date | None] = mapped_column(Date, nullable=True)
    value_numeric: Mapped[Decimal | None] = mapped_column(Numeric(20, 6), nullable=True)
    value_text: Mapped[str | None] = mapped_column(Text, nullable=True)
    value_json: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
    unit: Mapped[str | None] = mapped_column(String(50), nullable=True)  # "USD/MWh", "MW", "%"
    confidence: Mapped[float] = mapped_column(Numeric(3, 2), default=1.0, server_default="1.0")
    source_url: Mapped[str | None] = mapped_column(String(500), nullable=True)
    review_status: Mapped[str] = mapped_column(
        String(20), nullable=False, server_default="auto_accepted"
    )  # "pending_review" | "auto_accepted" | "approved" | "rejected"
    reviewed_by: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id", ondelete="SET NULL"), nullable=True
    )
    reviewed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

    __table_args__ = (
        Index("ix_market_data_processed_data_type", "data_type"),
        Index("ix_market_data_processed_effective_date", "effective_date"),
        Index("ix_market_data_processed_org", "org_id"),
        Index("ix_market_data_processed_review_status", "review_status"),
    )


class DataReviewQueue(BaseModel):
    """Human review queue for low-confidence or anomalous data points."""

    __tablename__ = "data_review_queue"

    org_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False)
    processed_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("market_data_processed.id", ondelete="CASCADE"),
        nullable=False,
    )
    assigned_to: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id", ondelete="SET NULL"), nullable=True
    )
    priority: Mapped[int] = mapped_column(Integer, default=0, server_default="0")
    reason: Mapped[str] = mapped_column(String(500), nullable=False)
    resolved_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

    __table_args__ = (
        Index("ix_data_review_queue_org", "org_id"),
        Index("ix_data_review_queue_processed", "processed_id"),
    )

File: apps/web/src/lib/market-enrichment.ts
"use client";

import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import { api } from "@/lib/api";

// ── Types ─────────────────────────────────────────────────────────────────────

export interface MarketDataSource {
  id: string;
  org_id: string;
  name: string;
  slug: string;
  description: string | null;
  source_type: "official_api" | "rss_feed" | "document" | "manual";
  tier: 1 | 2 | 3 | 4;
  base_url: string | null;
  legal_basis: "public_data" | "licensed" | "fair_use" | "manual_entry";
  rate_limit_per_hour: number;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface FetchLog {
  id: string;
  org_id: string;
  source_id: string;
  status: "pending" | "running" | "success" | "failed" | "rate_limited";
  records_fetched: number;
  records_new: number;
  records_updated: number;
  error_message: string | null;
  started_at: string | null;
  completed_at: string | null;
  created_at: string;
}

export interface MarketDataProcessed {
  id: string;
  org_id: string;
  raw_id: string | null;
  data_type: "price" | "policy" | "project_pipeline" | "macro_indicator" | "news";
  category: string;
  region: string | null;
  technology: string | null;
  effective_date: string | null;
  value_numeric: number | null;
  value_text: string | null;
  value_json: Record<string, unknown> | null;
  unit: string | null;
  confidence: number;
  source_url: string | null;
  review_status: "pending_review" | "auto_accepted" | "approved" | "rejected";
  reviewed_by: string | null;
  reviewed_at: string | null;
  created_at: string;
}

export interface ReviewQueueItem {
  id: string;
  org_id: string;
  processed_id: string;
  assigned_to: string | null;
  priority: number;
  reason: string;
  resolved_at: string | null;
  created_at: string;
  processed: MarketDataProcessed | null;
}

export interface MarketEnrichmentDashboard {
  sources_count: number;
  active_sources_count: number;
  records_today: number;
  pending_review_count: number;
  recent_fetches: FetchLog[];
}

export interface SourceFilters {
  tier?: number;
  is_active?: boolean;
}

export interface DataFilters {
  data_type?: string;
  category?: string;
  region?: string;
  technology?: string;
  effective_date_from?: string;
  effective_date_to?: string;
  review_status?: string;
  skip?: number;
  limit?: number;
}

// ── Query key factory ─────────────────────────────────────────────────────────

export const marketEnrichmentKeys = {
  all: ["market-enrichment"] as const,
  dashboard: () => [...marketEnrichmentKeys.all, "dashboard"] as const,
  sources: (filters?: SourceFilters) =>
    [...marketEnrichmentKeys.all, "sources", filters] as const,
  sourceLogs: (sourceId: string) =>
    [...marketEnrichmentKeys.all, "sources", sourceId, "logs"] as const,
  data: (filters?: DataFilters) => [...marketEnrichmentKeys.all, "data", filters] as const,
  dataRecord: (id: string) => [...marketEnrichmentKeys.all, "data", id] as const,
  reviewQueue: () => [...marketEnrichmentKeys.all, "review-queue"] as const,
};

// ── Hooks ─────────────────────────────────────────────────────────────────────

export function useMarketEnrichmentDashboard() {
  return useQuery<MarketEnrichmentDashboard>({
    queryKey: marketEnrichmentKeys.dashboard(),
    queryFn: () => api.get("/market-enrichment/dashboard").then((r) => r.data),
    staleTime: 60_000,
  });
}

export function useMarketDataSources(filters: SourceFilters = {}) {
  const params = new URLSearchParams();
  if (filters.tier != null) params.set("tier", String(filters.tier));
  if (filters.is_active != null) params.set("is_active", String(filters.is_active));

  return useQuery<MarketDataSource[]>({
    queryKey: marketEnrichmentKeys.sources(filters),
    queryFn: () =>
      api.get(`/market-enrichment/sources?${params}`).then((r) => r.data),
    staleTime: 120_000,
  });
}

export function useSourceFetchLogs(sourceId: string | null) {
  return useQuery<FetchLog[]>({
    queryKey: marketEnrichmentKeys.sourceLogs(sourceId ?? ""),
    queryFn: () =>
      api.get(`/market-enrichment/sources/${sourceId}/logs`).then((r) => r.data),
    enabled: !!sourceId,
  });
}

export function useMarketData(filters: DataFilters = {}) {
  const params = new URLSearchParams();
  if (filters.data_type) params.set("data_type", filters.data_type);
  if (filters.category) params.set("category", filters.category);
  if (filters.region) params.set("region", filters.region);
  if (filters.technology) params.set("technology", filters.technology);
  if (filters.effective_date_from) params.set("effective_date_from", filters.effective_date_from);
  if (filters.effective_date_to) params.set("effective_date_to", filters.effective_date_to);
  if (filters.review_status) params.set("review_status", filters.review_status);
  if (filters.skip != null) params.set("skip", String(filters.skip));
  if (filters.limit != null) params.set("limit", String(filters.limit));

  return useQuery<MarketDataProcessed[]>({
    queryKey: marketEnrichmentKeys.data(filters),
    queryFn: () =>
      api.get(`/market-enrichment/data?${params}`).then((r) => r.data),
    staleTime: 60_000,
  });
}

export function useReviewQueue() {
  return useQuery<ReviewQueueItem[]>({
    queryKey: marketEnrichmentKeys.reviewQueue(),
    queryFn: () => api.get("/market-enrichment/review-queue").then((r) => r.data),
    refetchInterval: 30_000,
  });
}

export function useCreateSource() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (body: Record<string, unknown>) =>
      api.post("/market-enrichment/sources", body).then((r) => r.data),
    onSuccess: () => qc.invalidateQueries({ queryKey: marketEnrichmentKeys.sources() }),
  });
}

export function useTriggerFetch() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (sourceId: string) =>
      api.post(`/market-enrichment/sources/${sourceId}/fetch`).then((r) => r.data),
    onSuccess: (_data, sourceId) => {
      qc.invalidateQueries({ queryKey: marketEnrichmentKeys.sourceLogs(sourceId) });
      qc.invalidateQueries({ queryKey: marketEnrichmentKeys.dashboard() });
    },
  });
}

export function useCreateManualEntry() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (body: Record<string, unknown>) =>
      api.post("/market-enrichment/data/manual", body).then((r) => r.data),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: marketEnrichmentKeys.data() });
      qc.invalidateQueries({ queryKey: marketEnrichmentKeys.dashboard() });
    },
  });
}

export function useReviewEntry() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: ({
      processedId,
      action,
      notes,
    }: {
      processedId: string;
      action: "approve" | "reject";
      notes?: string;
    }) =>
      api
        .post(`/market-enrichment/data/${processedId}/review`, { action, notes })
        .then((r) => r.data),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: marketEnrichmentKeys.reviewQueue() });
      qc.invalidateQueries({ queryKey: marketEnrichmentKeys.data() });
      qc.invalidateQueries({ queryKey: marketEnrichmentKeys.dashboard() });
    },
  });
}

// ── Constants ─────────────────────────────────────────────────────────────────

export const DATA_TYPES = [
  { value: "price", label: "Price" },
  { value: "policy", label: "Policy" },
  { value: "project_pipeline", label: "Project Pipeline" },
  { value: "macro_indicator", label: "Macro Indicator" },
  { value: "news", label: "News" },
];

export const TIER_LABELS: Record<number, string> = {
  1: "Official API",
  2: "RSS Feed",
  3: "Document",
  4: "Manual",
};

export const LEGAL_BASIS_LABELS: Record<string, string> = {
  public_data: "Public Data",
  licensed: "Licensed",
  fair_use: "Fair Use",
  manual_entry: "Manual Entry",
};

export const REVIEW_STATUS_BADGE: Record<string, string> = {
  pending_review: "bg-yellow-100 text-yellow-700",
  auto_accepted: "bg-blue-100 text-blue-700",
  approved: "bg-green-100 text-green-700",
  rejected: "bg-red-100 text-red-700",
};

export const FETCH_STATUS_BADGE: Record<string, string> = {
  pending: "bg-gray-100 text-gray-600",
  running: "bg-blue-100 text-blue-700",
  success: "bg-green-100 text-green-700",
  failed: "bg-red-100 text-red-700",
  rate_limited: "bg-orange-100 text-orange-700",
};
